### PR TITLE
fix wrong sequence of timetick stats on proxy #21855

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: #21855

### CODE REVIEW

File: internal/proxy/task_scheduler.go [hunk 1]
```diff
@@ -228,16 +228,22 @@ type dmTaskQueue struct {
 func (queue *dmTaskQueue) Enqueue(t task) error {
 	queue.statsLock.Lock()
 	defer queue.statsLock.Unlock()
-	err := queue.addPChanStats(t)
+	//1. preAdd will check whether provided task is valid or addable
+	//and get the current pChannels for this dmTask
+	pChannels, dmt, err := queue.preAddPChanStats(t)
```
Comment (czs007) [inline internal/proxy/task_scheduler.go:233]: preAddPChanStats just generate channels and dmlTask?

**Comment by MrPresent-Han** [comment]
rerun checker

File: internal/proxy/task_scheduler.go [hunk 2]
```diff
@@ -228,16 +228,22 @@ type dmTaskQueue struct {
...
 	}
+	//2. enqueue dml task
 	err = queue.baseTaskQueue.Enqueue(t)
 	if err != nil {
-		queue.popPChanStats(t)
 		return err
 	}
-
+	//3. if preAdd succeed, commit will use pChannels got previously when preAdding and will definitely succeed
+	queue.commitPChanStats(dmt, pChannels)
```
Comment (czs007) [inline internal/proxy/task_scheduler.go:243]: Will there be a problem with entering the queue first and then updating pChanStats?

File: internal/proxy/task_scheduler.go (hunk 1 above)
Comment (MrPresent-Han) [reply internal/proxy/task_scheduler.go:233]: to check whether all channels are available and the input task can be reflected to dmlTask
if these two conditions are not fulfilled, preAdd will fail and ensure the atomic property of the whole process

File: internal/proxy/task_scheduler.go (hunk 2 above)
Comment (MrPresent-Han) [reply internal/proxy/task_scheduler.go:243]: if ensue successfully but addPchanStats fail, there will an inconsistency

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**